File: java/GameAccountSimulator/src/gameAccountSimulator/LandCollection.java
package gameAccountSimulator;

import java.util.*;

public class LandCollection {

    LandCollection() {
        factory = new LandFactory();
        lands = new ArrayList<Land>();
        recentlyBought = new ArrayList<Land>();
    }

    public void createCommonLandCollection() {
        lands.add(factory.newLand(LandType.FARM));
        lands.add(factory.newLand(LandType.BARRAKS));
        lands.add(factory.newLand(LandType.BLACKSMITH));
        lands.add(factory.newLand(LandType.CASTLE_KEEP));
        lands.add(factory.newLand(LandType.LUMBER_MILL));
        lands.add(factory.newLand(LandType.ROYAL_CASTLE));
        lands.add(factory.newLand(LandType.ROYAL_TRADE_ROUTE));
        lands.add(factory.newLand(LandType.SHIPYARD));
        lands.add(factory.newLand(LandType.STONE_QUARRY));
        lands.add(factory.newLand(LandType.TAVERN));
        lands.add(factory.newLand(LandType.TEMPLE));
        lands.add(factory.newLand(LandType.VILLAGE));
        lands.add(factory.newLand(LandType.GOLD_MINE));
    }

    public void createCustomizedLandCollection(int [] counts) {
        for (int i = 0; i < counts.length; i++) {
            lands.add(factory.newLand(LandType.values()[i], counts[i]));
        }
    }

    public void printLandList() {

        for (Iterator iter = lands.iterator(); iter.hasNext();) {

            Land land = (Land) iter.next();

            System.out.println(land.toString());
        }

    }

    public Land getBestLand() {
        Collections.sort(lands, new LandIncomePerPriceComparator());
        
        return lands.get(0);
    }

    public double buyLand(Land land, double budget) {
        double balance = budget - land.getCurrentPrice();
        
        if (balance > 0.0) {
            land.incQuantity();
            recentlyBought.add(land);
        }

        return balance;
    }

    public double getTotalIncome() {
        double income = 0.0;

        for (Iterator iter = lands.iterator(); iter.hasNext();) {

            Land land = (Land) iter.next();

            income += land.getTotalIncome();
        }

        return income;
    }

    public Land getEffortableLand(double budget) {
        Land effortableLand = null;
        Collections.sort(lands, new LandCurrentPriceComparator());

        Iterator iter = lands.iterator();
        while (iter.hasNext()) {
            Land land = (Land) iter.next();
            if (land.getCurrentPrice() <= budget) {
                effortableLand = land;
                break;
            }

        }

        return effortableLand;
    }

    public void printRecentlyBought() {
        Collections.sort(recentlyBought, new LandCurrentPriceComparator());
        for (Land l : recentlyBought) {
            System.out.println(String.format("Bought land! %s", l.toString()));
        }
        System.out.println("-------------------------------------------------");
        recentlyBought.clear();
    }
    
    private ArrayList<Land> lands;
    private LandFactory factory;
    private ArrayList<Land> recentlyBought;

}

File: objc/KingdomsSimulator/KingdomsSimulator/LandCollection.m
//
//  LandCollection.m
//  KingdomsSimulator
//
//  Created by Merxbauer Jaroslav on 02.10.11.
//  Copyright 2011 __MyCompanyName__. All rights reserved.
//

#import "LandCollection.h"

@implementation LandCollection

- (id)init
{
    self = [super init];
    if (self) {
        factory = [[LandFactory alloc] init];
        lands = [[NSMutableArray alloc] init];
        receipt = [[PurchaseInfo alloc] init];
    }
    
    return self;
}

- (void) dealloc {
    [factory release];
    [lands release];
    [receipt release];
}

- (void) createCommonLandCollection {
    [lands addObject:[factory newLandFromType:FARM]];
    [lands addObject:[factory newLandFromType:BARRAKS]];
    [lands addObject:[factory newLandFromType:BLACKSMITH]];
    [lands addObject:[factory newLandFromType:CASTLE_KEEP]];
    [lands addObject:[factory newLandFromType:LUMBER_MILL]];
    [lands addObject:[factory newLandFromType:ROYAL_CASTLE]];
    [lands addObject:[factory newLandFromType:ROYAL_TRADE_ROUTE]];
    [lands addObject:[factory newLandFromType:SHIPYARD]];
    [lands addObject:[factory newLandFromType:STONE_QUARRY]];
    [lands addObject:[factory newLandFromType:TAVERN]];
    [lands addObject:[factory newLandFromType:TEMPLE]];
    [lands addObject:[factory newLandFromType:VILLAGE]];
    [lands addObject:[factory newLandFromType:GOLD_MINE]];
}

- (void) createLandCollectionFromArraySetup:(NSArray *)arraySetup {
    for (int i = 0; i < [arraySetup count]; i++) {
        NSNumber* count = (NSNumber*) [arraySetup objectAtIndex:i];
        [lands addObject:[factory newLandFromType:i quantity:[count intValue]]];
    }
}

- (void) printLandList {
    for (Land* land in lands) {
        printf("%s", [[land description] UTF8String]);
    }
}

- (Land*) getBestLand {
    [lands sortUsingSelector:@selector(compareLandBasedOnIncomePerPrice:)];
    return [lands objectAtIndex:0];
}

- (double) buyLand: (Land*) land withAvailableBudget: (double) budget {
    double balance = budget - [land getCurrentPrice];
    if (balance > 0.0) {
        [receipt purchaseLand: land];
        [land incQuantity];
    }
    return balance;
}

- (double) getTotalIncome {
    double income = 0.0;
    for (Land* land in lands) {
        income += [land getTotalIncome];
    }
    return income;
}

- (Land*) getEffortableLandWithBudget: (double) budget {
    Land* effortableLand = nil;
    [lands sortUsingSelector:@selector(compareLandBasedOnCurrentPrice:)];
    for (Land* land in lands) {
        if ([land getCurrentPrice] <= budget) {
            effortableLand = land;
            break;
        }
    }
    return effortableLand;
}

- (void) printRecentlyBought {
    [receipt print];
    [receipt clear];
    printf("--------------------------------------------------------------------------------\n");
}

@end
